File: 064. Minimum Path Sum/dp.cpp
class Solution {
public:
    int ret = INT_MAX;
    int minPathSum(vector<vector<int>>& grid) {
        if(grid.size() == 0 || grid[0].size() == 0)
            return 0;
        int row = grid.size(), col = grid[0].size();
        //vector<vector<int>> dp(row, vector<int>(col, 0));
        for(int i = 1; i < col; ++i){
            grid[0][i] = grid[0][i] + grid[0][i - 1];
        }
        for(int i = 1; i < row; ++i){
            grid[i][0] = grid[i][0] + grid[i - 1][0];
        }
        
        for(int i = 1; i < row; ++i){
            for(int j = 1; j < col; ++j){
                grid[i][j] += min(grid[i - 1][j], grid[i][j - 1]);
            }
        }
        
        return grid[row - 1][col - 1];
    }

};

File: 0799. Champagne Tower/dp.java
class Solution {
    public double champagneTower(int n, int row, int index) {
        double[][] dp = new double[101][101];
        for(int i = 0; i < 101; ++i){
            for(int j = 0; j < 101; ++j){
                dp[i][j] = 0.0;
            }
        }
        dp[0][0] = n;
        
        for(int i = 0; i < 100; ++i){
            for(int j = 0; j <= i; ++j){
                if(dp[i][j] > 1){
                    double remain = dp[i][j] - 1;
                    dp[i + 1][j] += remain / 2;
                    dp[i + 1][j + 1] += remain / 2;
                    dp[i][j] = 1;
                }
            }
        }
        return dp[row][index];
    }
}